File: app/src/main/java/com/example/neon/deardiary/diarylist/DiaryListPresenter.java
package com.example.neon.deardiary.diarylist;

import android.util.Log;

import com.example.neon.deardiary.data.Diary;
import com.example.neon.deardiary.data.source.DataSource;
import com.example.neon.deardiary.data.source.local.originaldb.DiaryLocalDataSource;

import java.util.ArrayList;
import java.util.Calendar;


class DiaryListPresenter implements DiaryListContract.Presenter {
    private static final String TAG = "DiaryListPresenter";
    private DataSource mDataSource;
    private DiaryListContract.View mView;

    DiaryListPresenter(DiaryLocalDataSource dataSource, DiaryListContract.View view) {
        mDataSource = dataSource;
        mView = view;
        mView.setPresenter(this);
    }

    @Override
    public void start() {
        loadDiaries(Calendar.getInstance());
    }

    public void loadDiaries(Calendar c) {
        mDataSource.queryAndAddDefault(c, new DataSource.LoadDiaryCallBack() {
            @Override
            public void onDiaryLoaded(ArrayList<Diary> diaryList) {
                mView.showDiary(diaryList);
            }

            @Override
            public void onDataNotAvailable() {
                Log.d(TAG, "onDataNotAvailable: 数据读取失败");

            }
        });

    }

    @Override
    public Diary getDiary(Calendar c) {
        final Diary[] diary = new Diary[1];
        mDataSource.queryByDay(c, new DataSource.GetDiaryCallBack() {
            @Override
            public void onDiaryGot(Diary foundDiary) {
                diary[0] = foundDiary;
            }

            @Override
            public void onDataNotAvailable() {
                diary[0] = null;
            }
        });
        return diary[0];

    }

    @Override
    public void insertDiary(Diary diary) {
        mDataSource.insertDiary(diary);
    }
}

File: app/src/main/java/com/example/neon/deardiary/diarylist/DiaryListPresenter.kt
package com.example.neon.deardiary.diarylist

import com.example.neon.deardiary.data.DataSource
import com.example.neon.deardiary.data.Diary
import com.example.neon.deardiary.data.LocalDataSource
import com.example.neon.deardiary.utils.LogUtil
import java.util.ArrayList
import java.util.Calendar


internal class DiaryListPresenter(dataSource: LocalDataSource, private val mView: DiaryListContract.View) : DiaryListContract.Presenter {
    private val mDataSource: DataSource

    init {
        mDataSource = dataSource
        mView.setPresenter(this)
    }

    override fun loadDiaries(c: Calendar) {
        mDataSource.queryAndAddDefault(c, object : DataSource.LoadDiariesCallBack {
            override fun onDiaryLoaded(diaryList: ArrayList<Diary>) {
                mView.showDiary(diaryList)
            }

            override fun onDataNotAvailable() {
                LogUtil.e(TAG,"onDataNotAvailable: 数据读取失败")
            }
        })

    }

    override fun getDiary(c: Calendar): Diary? {
        var foundDiary: Diary? = null
        mDataSource.queryByDay(c, object : DataSource.LoadDiaryCallBack {
            override fun onDiaryGot(diary: Diary) {
                foundDiary = diary
            }

            override fun onDataNotAvailable() {
                foundDiary = null
            }
        })
        return foundDiary

    }

    override fun insertDiary(diary: Diary) {
        mDataSource.insertDiary(diary)
    }

    companion object {
        private val TAG = "DiaryListPresenter"
    }
}
